crates/bevy_reflect/bevy_reflect_derive/src/impls/enums.rs

```diff
@@ -262,126 +261,91 @@ struct EnumImpls {
 fn generate_impls(reflect_enum: &ReflectEnum, ref_index: &Ident, ref_name: &Ident) -> EnumImpls {
     let bevy_reflect_path = reflect_enum.meta().bevy_reflect_path();
 
-    let mut variant_info: Vec<proc_macro2::TokenStream> = Vec::new();
-    let mut enum_field: Vec<proc_macro2::TokenStream> = Vec::new();
-    let mut enum_field_at: Vec<proc_macro2::TokenStream> = Vec::new();
-    let mut enum_index_of: Vec<proc_macro2::TokenStream> = Vec::new();
-    let mut enum_name_at: Vec<proc_macro2::TokenStream> = Vec::new();
-    let mut enum_field_len: Vec<proc_macro2::TokenStream> = Vec::new();
-    let mut enum_variant_name: Vec<proc_macro2::TokenStream> = Vec::new();
-    let mut enum_variant_type: Vec<proc_macro2::TokenStream> = Vec::new();
+    let mut variant_info = Vec::new();
+    let mut enum_field = Vec::new();
+    let mut enum_field_at = Vec::new();
+    let mut enum_index_of = Vec::new();
+    let mut enum_name_at = Vec::new();
+    let mut enum_field_len = Vec::new();
+    let mut enum_variant_name = Vec::new();
+    let mut enum_variant_type = Vec::new();
 
     for variant in reflect_enum.active_variants() {
         let ident = &variant.data.ident;
         let name = ident.to_string();
         let unit = reflect_enum.get_unit(ident);
 
-        match &variant.fields {
-            EnumVariantFields::Unit => {
-                variant_info.push(quote! {
-                    #bevy_reflect_path::VariantInfo::Unit(
-                        #bevy_reflect_path::UnitVariantInfo::new(#name)
-                    )
-                });
-                enum_variant_name.push(quote! {
-                    #unit => #name
-                });
-                enum_variant_type.push(quote! {
-                    #unit => #bevy_reflect_path::VariantType::Unit
-                });
-            }
-            EnumVariantFields::Unnamed(fields) => {
+        // This is equivalent to a |fields: &Vec<StructField>, to_run: |usize, usize, &StructField| -> TokenStream|
+        // closure. Issue is that the closure cannot itself accept another closure, because closures cannot vary in
+        // the concrete type they accept, and each closure has a different concrete type.
+        macro_rules! for_fields {
```

Comment:
I managed to get this to work with a function like:
```rust
fn for_fields<F>(fields: &[StructField], mut field_generator: F) -> (usize, Vec<proc_macro2::TokenStream>)
where
  F: FnMut(usize, usize, &StructField) -> proc_macro2::TokenStream,
{
  let mut field_info = Vec::new();
  let mut reflect_idx: usize = 0;
  for (declar_field, field) in fields.iter().enumerate() {
    if field.attrs.ignore {
      // Ignored field
      continue;
    }
    field_info.push(field_generator(reflect_idx, declar_field, field));
    reflect_idx += 1;
  }
  (reflect_idx, field_info)
}
```

Reply:
Actually awesome. I'm so sad closures are not expressive enough to do that kind of things.